src/adapter/src/coord/introspection.rs

```diff
@@ -0,0 +1,191 @@
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0.
+
+//! Special cases related to the "introspection" of Materialize
+//!
+//! Every Materialize deployment has a pre-installed [`mz_introspection`] cluster, which
+//! has several indexes to speed up common introspection queries. We also have a special
+//! `mz_introspection` role, which can be used by support teams to diagnose a deployment.
+//! For each of these use cases, we have some special restrictions we want to apply. The
+//! logic around these restrictions is defined here.
+//!
+//!
+//! [`mz_introspection`]: https://materialize.com/docs/sql/show-clusters/#mz_introspection-system-cluster
+
+use mz_ore::collections::HashSet;
+use mz_repr::GlobalId;
+use mz_sql::catalog::SessionCatalog;
+use mz_sql::plan::Plan;
+use once_cell::sync::Lazy;
+
+use crate::catalog::{Catalog, Cluster};
+use crate::notice::AdapterNotice;
+use crate::rbac;
+use crate::session::Session;
+
+use crate::{
+    catalog::builtin::{
+        INFORMATION_SCHEMA, MZ_CATALOG_SCHEMA, MZ_INTERNAL_SCHEMA, MZ_INTROSPECTION_CLUSTER,
+        MZ_INTROSPECTION_ROLE, PG_CATALOG_SCHEMA,
+    },
+    AdapterError,
+};
+
+/// The schema's a user is allowed to query from the `mz_introspection` cluster
+static ALLOWED_SCHEMAS: Lazy<HashSet<&str>> = Lazy::new(|| {
+    HashSet::from([
+        MZ_CATALOG_SCHEMA,
+        PG_CATALOG_SCHEMA,
+        MZ_INTERNAL_SCHEMA,
+        INFORMATION_SCHEMA,
+    ])
+});
+
+/// Checks whether or not we should automatically run a query on the `mz_introspection`
+/// cluster, as opposed to whatever the current default cluster is.
+pub fn auto_run_on_introspection<'a, 's>(
+    catalog: &'a Catalog,
+    session: &'s mut Session,
+    depends_on: impl IntoIterator<Item = GlobalId>,
+) -> Option<&'a Cluster> {
+    // Allow users to opt out of this behavior.
+    if !session.vars().force_introspection_cluster() {
+        return None;
+    }
+
+    // Make sure we only depend on the system catalog.
+    let allowed = depends_on.into_iter().all(|id| {
+        let entry = catalog.get_entry(&id);
+        let full_name = catalog.resolve_full_name(entry.name(), Some(session.conn_id()));
+        ALLOWED_SCHEMAS.contains(full_name.schema.as_str())
+    });
+
+    // If we're allowed to run on the mz_introspection cluster, make sure we can resolve it.
+    if allowed {
+        let intros_cluster = catalog.resolve_cluster(MZ_INTROSPECTION_CLUSTER.name);
+        let active_cluster = catalog.active_cluster(session);
+
+        match (intros_cluster, active_cluster) {
+            (Ok(intros), active) => {
+                tracing::info!("Running on mz_introspection cluster");
+                // If we're running on a different cluster than the active
+                // one, notify the user.
+                if active.map(|c| c.id != intros.id).unwrap_or(true) {
+                    session.add_notice(AdapterNotice::AutoRunOnIntrospectionCluster);
```

Comment:
I suspect this will likely be too spammy (every `SHOW` command will cause this to be printed out, right?) but no harm in trying it out and it will be easy to reverse.

Reply:
Oh neat, I don't think we really take advantage of the fact that we can deliver notices at the debug level.